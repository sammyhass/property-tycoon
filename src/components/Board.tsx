import { AspectRatio, Box, Flex, Text } from '@chakra-ui/react';
import React from 'react';

const CELL_SIZE = 80;

const NUM_ROWS = 10;
const NUM_COLS = 10;

export default function Board() {
  return (
    <Box width={`${NUM_COLS * CELL_SIZE}px`}>
      <FilledRow row={0} />
      {new Array(NUM_ROWS - 2).fill(0).map((_, i) => (
        <SpacedRow row={i + 1} />
      ))}
      <FilledRow row={9} />
    </Box>
  );
}

const BoardCell = ({ n }: { n: number }) => {
  return (
    <AspectRatio
      ratio={1}
      minW={CELL_SIZE}
      h={CELL_SIZE}
      bg="blue"
      onClick={() => alert(n)}
    >
      <Flex>
        <Text>Board</Text>
      </Flex>
    </AspectRatio>
  );
};

const FilledRow = ({ row }: { row: number }) => {
  return (
    <Flex>
      {new Array(NUM_COLS).fill(0).map((_, i) => (
        <BoardCell
          n={row === 0 ? i : NUM_COLS + NUM_COLS - 3 + (NUM_COLS - i)}
        />
      ))}
    </Flex>
  );
};

const SpacedRow = ({ row }: { row: number }) => {
  return (
    <Flex justifyContent={'space-between'}>
      <BoardCell n={3 * NUM_COLS - 3 + (NUM_ROWS - 1 - row)} />
      <BoardCell n={row + NUM_COLS - 1} />
    </Flex>
  );
};
