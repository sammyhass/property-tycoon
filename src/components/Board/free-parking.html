<div class="board-space square free-parking">
	<div class="rotate top-right">
		Free
		<img border="0" draggable="false"/>
		Parking
	</div>
</div>