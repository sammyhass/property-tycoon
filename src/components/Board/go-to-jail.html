<div class="boardSpace square goToJail">
	<div class="rotate bottomRight">
		Go to
		<img border="0" draggable="false"/>
		Jail
	</div>
</div>