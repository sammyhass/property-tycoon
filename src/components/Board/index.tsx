import { GameContextT } from '@/hooks/useGameContext';
import { Box } from '@chakra-ui/react';
import BoardSpace from './spaces';

export default function GameBoard({
  settings,
}: {
  settings: GameContextT['gameSettings'];
}) {
  return (
    <Box>
      {settings?.BoardSpaces.sort(
        (a, b) => a.board_position - b.board_position
      ).map(space => (
        <BoardSpace
          key={space.board_position}
          property={
            settings?.Properties.find(p => p.id === space.property_id) ?? null
          }
          {...space}
        />
      ))}
    </Box>
  );
}
