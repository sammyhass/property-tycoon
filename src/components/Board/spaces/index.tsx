import { formatPrice } from '@/util/formatPrice';
import { propertyGroupToCSS } from '@/util/property-colors';
import { Box, Text } from '@chakra-ui/react';
import {
  BoardSpace as BoardSpaceT,
  CardType,
  GameProperty,
} from '@prisma/client';
import React from 'react';
import styles from './spaces.module.css';

type BoardSpaceProps = BoardSpaceT & { property: GameProperty | null };

export default function BoardSpace(props: BoardSpaceProps) {
  switch (props.space_type) {
    case 'PROPERTY':
      return <BoardSpace.Property property={props.property} />;
    case 'GO':
      return <BoardSpace.Go />;
    case 'TAKE_CARD':
      return <BoardSpace.TakeCard cardType={props?.take_card ?? 'POT_LUCK'} />;
    case 'GO_TO_JAIL':
      return <BoardSpace.GoToJail />;
    case 'JUST_VISIT':
      return <BoardSpace.Jail />;
    case 'FREE_PARKING':
      return <BoardSpace.FreeParking />;
    default:
      return <BoardSpace.Empty />;
  }
}

BoardSpace.Property = ({ property }: { property: GameProperty | null }) => (
  <div
    className={`${styles.boardSpace} ${styles.property}`}
    style={{
      background: propertyGroupToCSS[property?.property_group_color ?? 'NONE'],
    }}
  >
    <div className={styles.boardBackground} />
    <div className={styles.propertyContent}>
      <div className={`${styles.propertyText}`}>{property?.name}</div>
      <div className={styles.propertyPrice}>
        {formatPrice(property?.price ?? 0)}
      </div>
    </div>
  </div>
);

BoardSpace.Empty = () => (
  <div className={styles.boardSpace}>
    <div className={styles.boardBackground} />
    <div>
      <Text fontSize="sm" p="0" m="0">
        Empty
      </Text>
    </div>
  </div>
);

/**
 * 'Special' board spaces such as tax/card pickup/properties whose group isn't a color (e.g. Stations/Utilities)
 */
BoardSpace.Special = ({
  title,
  imageComponent: ImageComponent,
}: {
  title: string;
  imageComponent: React.FC;
}) => (
  <div className={`${styles.boardSpace} ${styles.special}`}>
    <div className={styles.specialText}>{title}</div>
    <ImageComponent />
  </div>
);

BoardSpace.Jail = () => (
  <div className={`${styles.boardSpace} ${styles.square} ${styles.inJail}`}>
    <div className={styles.just}>Just</div>
    <div className={styles.visiting}>Visiting</div>
    <div className={styles.inner}>
      <div>
        In
        <Box>
          <Text fontSize="6xl">‚õìÔ∏è</Text>
        </Box>
        Jail
      </div>
    </div>
  </div>
);

BoardSpace.GoToJail = () => (
  <div className={`${styles.boardSpace} ${styles.square} ${styles.goToJail}`}>
    <div className="rotate bottomRight">
      Go to
      <Box>
        <Text fontSize={'6xl'}>ü§¶</Text>
      </Box>
      Jail
    </div>
  </div>
);

BoardSpace.Go = () => {
  return (
    <div className={`${styles.boardSpace} ${styles.square} ${styles.go}`}>
      <svg className="arrow" viewBox="0 0 14 70" fill="#AB3126" stroke="#000">
        <polygon points="0.861 12.096 7.133 1.054 13.214 11.969 9.32 11.969 9.32 56.581 12.467 59.728 12.467 68.746 6.87 63.66 1.582 69.057 1.471 59.688 4.817 56.564 4.817 12.061" />
      </svg>
      <div className={`${styles.rotate} ${styles.bottomLeft}`}>
        <div className={styles.goText}>Collect ¬£200 salary as you pass</div>
        <div className={styles.goTextBig}>GO</div>
      </div>
    </div>
  );
};

BoardSpace.FreeParking = () => (
  <div
    className={`${styles.boardSpace} ${styles.square} ${styles.freeParking}`}
  >
    <div className={`${styles.rotate} ${styles.topRight}`}>
      Free
      <Box>
        <Text fontSize={'6xl'}>üéÅ</Text>
      </Box>
      Parking
    </div>
  </div>
);

BoardSpace.TakeCard = ({ cardType }: { cardType: CardType }) => (
  <BoardSpace.Special
    title={cardType === 'POT_LUCK' ? 'Pot Luck' : 'Opportunity Knocks'}
    imageComponent={() => (
      <Text fontSize={'6xl'}>{cardType === 'POT_LUCK' ? 'üé≤' : 'üî•'}</Text>
    )}
  />
);
