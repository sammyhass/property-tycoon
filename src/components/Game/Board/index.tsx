import { GameContextT, TokenType } from '@/hooks/useGameContext';
import { Flex } from '@chakra-ui/react';
import BoardSpace from './spaces';

// Gameboard with players on the board
export interface GameBoardProps {
  settings: GameContextT['gameSettings'];
  positions: { [key in keyof GameContextT['state']]: number };
}

export default function GameBoard({ positions, settings }: GameBoardProps) {
  return (
    <Flex wrap="wrap" transformOrigin="0 0">
      {(settings?.BoardSpaces ?? [])
        .sort((a, b) => a.board_position - b.board_position)
        .map(space => (
          <BoardSpace
            hasPlayers={Object.keys(positions ?? {})
              .filter(
                key => positions[key as TokenType] === space.board_position
              )
              .map(v => v as TokenType)}
            key={space.board_position}
            property={
              settings?.Properties.find(p => p.id === space.property_id) ?? null
            }
            {...space}
          />
        ))}
    </Flex>
  );
}
