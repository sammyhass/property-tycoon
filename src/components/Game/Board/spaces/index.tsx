import { TokenType } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import { propertyGroupToCSS } from '@/util/property-colors';
import { Box, ChakraProps, Flex, Text } from '@chakra-ui/react';
import { faFaucet, faTrain } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  BoardSpace as BoardSpaceT,
  CardType,
  GameProperty,
} from '@prisma/client';
import React from 'react';
import BoardSpacePlayers from './players';
import styles from './spaces.module.scss';

// Board Space Aspect ratio as defined in the css file. Useful to use when we need our spaces to show larger or smaller
export const BOARD_SPACE_ASPECT_RATIO = 11 / 7;

// Has player can be used to show the player's token on the board
type HasPlayerT = { hasPlayers?: TokenType[] };

type BoardSpaceProps = BoardSpaceT & {
  property: GameProperty | null;

  // Whether or not to rotate this board space 90 degrees
};

export default function BoardSpace(props: BoardSpaceProps & HasPlayerT) {
  return (
    <Box pos={'relative'}>
      <BoardSpacePlayers players={props.hasPlayers ?? []} />
      <BoardSpaceInner {...props} />
    </Box>
  );
}
// Board Spaces on the board. There are different types of board spaces, so
// so we need to determine which type of board space we are rendering.
const BoardSpaceInner = (props: BoardSpaceProps) => {
  switch (props.space_type) {
    case 'PROPERTY':
      return <BoardSpace.Property property={props.property} />;
    case 'GO':
      return <BoardSpace.Go />;
    case 'TAKE_CARD':
      return <BoardSpace.TakeCard cardType={props?.take_card ?? 'POT_LUCK'} />;
    case 'GO_TO_JAIL':
      return <BoardSpace.GoToJail />;
    case 'JUST_VISIT':
      return <BoardSpace.Jail />;
    case 'FREE_PARKING':
      return <BoardSpace.FreeParking />;
    case 'TAX':
      return <BoardSpace.Tax taxCost={props.tax_cost ?? 0} />;
    default:
      return <BoardSpace.Empty />;
  }
};

BoardSpace.Property = ({
  property,
  ...props
}: { property: GameProperty | null } & HasPlayerT & ChakraProps) => (
  <Box
    className={`${styles.boardSpace} ${styles.property}`}
    style={{
      background: propertyGroupToCSS[property?.property_group_color ?? 'NONE'],
    }}
    {...props}
  >
    <div className={styles.boardBackground} />
    <div className={styles.propertyContent}>
      <Flex w="100%" justify="center" align="center">
        {property?.property_group_color === 'STATION' ? (
          <FontAwesomeIcon icon={faTrain} size="2x" />
        ) : property?.property_group_color === 'UTILITIES' ? (
          <Flex w="100%" justify="center" align="center">
            <FontAwesomeIcon icon={faFaucet} size="2x" />
          </Flex>
        ) : (
          <></>
        )}
      </Flex>

      <div className={`${styles.propertyText}`}>{property?.name}</div>
      <div className={styles.propertyPrice}>
        {formatPrice(property?.price ?? 0)}
      </div>
    </div>
  </Box>
);

BoardSpace.Empty = (props: HasPlayerT) => (
  <div className={styles.boardSpace}>
    <div className={styles.boardBackground} />
    <div>
      <Text fontSize="sm" p="0" m="0">
        Empty
      </Text>
    </div>
  </div>
);

/**
 * 'Special' board spaces such as tax/card pickup/properties whose group isn't a color (e.g. Stations/Utilities)
 */
BoardSpace.Special = ({
  title,
  imageComponent: ImageComponent,
}: {
  title: string;
  imageComponent: React.FC;
} & HasPlayerT) => (
  <div className={`${styles.boardSpace} ${styles.special}`}>
    <div className={styles.specialText}>{title}</div>
    <ImageComponent />
  </div>
);

BoardSpace.Jail = (props: { jail?: HasPlayerT; visit?: HasPlayerT }) => (
  <div className={`${styles.boardSpace} ${styles.square} ${styles.inJail}`}>
    <div className={styles.just}>Just</div>
    <div className={styles.visiting}>Visiting</div>
    <div className={styles.inner}>
      <div>
        In
        <Box>
          <Text fontSize="6xl">‚õìÔ∏è</Text>
        </Box>
        Jail
      </div>
    </div>
  </div>
);

BoardSpace.GoToJail = (props: HasPlayerT) => (
  <div className={`${styles.boardSpace} ${styles.square} ${styles.goToJail}`}>
    <div className={`${styles.rotate} ${styles.bottomRight}`}>
      Go to
      <Box>
        <Text fontSize={'6xl'}>ü§¶</Text>
      </Box>
      Jail
    </div>
  </div>
);

BoardSpace.Go = (props: HasPlayerT) => {
  return (
    <div className={`${styles.boardSpace} ${styles.square} ${styles.go}`}>
      <svg className="arrow" viewBox="0 0 14 70" fill="#AB3126" stroke="#000">
        <polygon points="0.861 12.096 7.133 1.054 13.214 11.969 9.32 11.969 9.32 56.581 12.467 59.728 12.467 68.746 6.87 63.66 1.582 69.057 1.471 59.688 4.817 56.564 4.817 12.061" />
      </svg>
      <div className={`${styles.rotate} ${styles.bottomLeft}`}>
        <div className={styles.goText}>Collect ¬£200 salary as you pass</div>
        <div className={styles.goTextBig}>GO</div>
      </div>
    </div>
  );
};

BoardSpace.FreeParking = () => (
  <div
    className={`${styles.boardSpace} ${styles.square} ${styles.freeParking}`}
  >
    <div className={`${styles.rotate} ${styles.topRight}`}>
      Free
      <Box>
        <Text fontSize={'6xl'}>üéÅ</Text>
      </Box>
      Parking
    </div>
  </div>
);

BoardSpace.TakeCard = ({ cardType }: { cardType: CardType }) => (
  <BoardSpace.Special
    title={cardType === 'POT_LUCK' ? 'Pot Luck' : 'Opportunity Knocks'}
    imageComponent={() => (
      <Text fontSize={'6xl'}>{cardType === 'POT_LUCK' ? 'üé≤' : 'üî•'}</Text>
    )}
  />
);

BoardSpace.Tax = ({ taxCost }: { taxCost: number }) => (
  <BoardSpace.Special
    title={`¬£${taxCost} Tax`}
    imageComponent={() => <Text fontSize={'6xl'}>üí∏</Text>}
  />
);
