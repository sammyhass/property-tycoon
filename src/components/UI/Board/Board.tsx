import { Box } from '@chakra-ui/react';
import React, { createContext, useContext } from 'react';
import BoardCell from './BoardCell';

const CELL_SIZE = 80;

const BOARD_SIZE = 10;

interface BoardProps {
  onTileClick?: (n: number) => void;
  boardSize: number;
  cellSize: number;
}

const BoardPropsContext = createContext<BoardProps>({
  boardSize: BOARD_SIZE,
  cellSize: CELL_SIZE,
});

export const useBoardProps = () => useContext(BoardPropsContext);

export default function Board({
  onTileClick,
  boardSize = BOARD_SIZE,
  cellSize = CELL_SIZE,
}: BoardProps) {
  return (
    <BoardPropsContext.Provider value={{ onTileClick, boardSize, cellSize }}>
      <Box
        sx={{
          display: 'grid',
          gridTemplateColumns: `repeat(${boardSize}, ${cellSize}px)`,
          gridTemplateRows: `repeat(${boardSize}, ${cellSize}px)`,
          gridGap: 3,
        }}
      >
        {/* Dummy element to space out center of grid */}
        <Box
          gridRowStart={2}
          gridRowEnd={boardSize - 2}
          gridColumnStart={2}
          gridColumnEnd={boardSize - 2}
        />
        {new Array(boardSize * 4 - 4).fill(0).map((_, i) => (
          <BoardCell n={i} key={i} />
        ))}
      </Box>
    </BoardPropsContext.Provider>
  );
}
