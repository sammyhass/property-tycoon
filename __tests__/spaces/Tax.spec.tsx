import { BoardSpaceTax } from '@/components/Game/Board/spaces';
import { render } from '@testing-library/react';
import React from 'react';

describe('Tax Board Space', () => {
  it('matches snapshot', () => {
    const { container } = render(<BoardSpaceTax taxCost={100} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the correct text', () => {
    const { container } = render(<BoardSpaceTax taxCost={100} />);

    expect(container.textContent).toContain('Â£100 Tax');
  });
});
