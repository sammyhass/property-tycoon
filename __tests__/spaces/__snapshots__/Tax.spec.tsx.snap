// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tax Board Space matches snapshot 1`] = `
<div
  class="boardSpace special"
>
  <div
    class="specialText"
  >
    Â£100 Tax
  </div>
  <p
    class="chakra-text css-n2bes8"
  >
    ðŸ’¸
  </p>
</div>
`;
